curvefs/src/client/block_device_client.h

```diff
@@ -25,80 +25,148 @@
+
+    /**
+     * @brief Get file status
+     * @param[in] filename
+     * @param[in] owner owner for filename
+     * @param[out] statInfo the struct for file status
+     * @return error code (CURVEFS_ERROR:*)
+     */
+    virtual CURVEFS_ERROR Stat(const std::string& filename,
+                               const std::string& owner,
+                               BlockDeviceStat* statInfo) = 0;
+
+    /**
+     * @brief Read from fd which init by Open()
+     * @param[in] buf read buffer
+     * @param[in] offset read start offset
+     * @param[in] length read length
+     * @return error code (CURVEFS_ERROR:*)
+     * @note the offset and length maybe not aligned
+     */
+    virtual CURVEFS_ERROR Read(char* buf, off_t offset, size_t length) = 0;
 
     /**
-     * @brief write
-     *
-     * @param fd file descriptor
-     * @param buf write buffer
-     * @param offset write start offset
-     * @param length write length
-     *
-     * @return error code
+     * @brief Write to fd which init by Open()
+     * @param[in] buf write buffer
+     * @param[in] offset write start offset
+     * @param[in] length write length
+     * @return error code (CURVEFS_ERROR:*)
+     * @note the offset and length maybe not aligned
      */
     virtual CURVEFS_ERROR Write(
         const char* buf, off_t offset, size_t length) = 0;
 };
 
 class BlockDeviceClientImpl : public BlockDeviceClient {
  public:
-    explicit BlockDeviceClientImpl(std::shared_ptr<FileClient> fileClient) :
-        fileClient_(fileClient) {}
-    virtual ~BlockDeviceClientImpl() {}
+    explicit BlockDeviceClientImpl(std::shared_ptr<FileClient> fileClient)
+        : fileClient_(fileClient) {}
 
-    CURVEFS_ERROR Init(const BlockDeviceClientOptions &options) override;
+    BlockDeviceClientImpl() = default;
 
-    CURVEFS_ERROR UnInit() override;
+    ~BlockDeviceClientImpl() {}
+
+    CURVEFS_ERROR Init(const BlockDeviceClientOptions& options) override;
+
+    void UnInit() override;
+
+    CURVEFS_ERROR Open(const std::string& filename,
+                       const std::string& owner) override;
+
+    CURVEFS_ERROR Close() override;
+
+    CURVEFS_ERROR Stat(const std::string& filename,
+                       const std::string& owner,
+                       BlockDeviceStat* statInfo) override;
 
     CURVEFS_ERROR Read(char* buf, off_t offset, size_t length) override;
 
     CURVEFS_ERROR Write(const char* buf, off_t offset, size_t length) override;
 
  private:
-    std::shared_ptr<FileClient> fileClient_;
-    bool isOpen_;
+    CURVEFS_ERROR WritePadding(char* writeBuffer, off_t writeStart,
+                               off_t writeEnd, off_t offset, size_t length);
+
+    CURVEFS_ERROR AlignRead(char* buf, off_t offset, size_t length);
+
+    CURVEFS_ERROR AlignWrite(const char* buf, off_t offset, size_t length);
+
+    bool IsAligned(off_t offset, size_t length);
+
+    Range CalcAlignRange(off_t start, off_t end);
+
+ private:
     uint64_t fd_;
```

Comment:
Can fd be equal to 0?,  if deed, maybe use int64_t as fd will be better, thus  can use -1 as invalid fd.

Reply:
> Can fd be equal to 0?, if deed, maybe use int as fd will be better, thus can use -1 as invalid fd.

done